src/Interpreters/ConcurrentHashJoin.cpp

```diff
@@ -620,6 +695,88 @@ void ConcurrentHashJoin::onBuildPhaseFinish()
                 },
                 getData(hash_joins[0])->maps.at(0));
         }
+
+        /// rebuild per-slot right-side nullmaps into slot 0 so that
+        /// non-joined rows saved due to NULL keys or ON-filtered rows are emitted only once
+        if (std::any_of(hash_joins.begin(), hash_joins.end(),
+                        [&](const auto &hj){ return !getData(hj)->nullmaps.empty(); }))
+        {
+            /// Keep the pointers in NullMapHolder to original ScatteredColumns, which remain valid
+            /// in their respective slots
+            HashJoin::NullmapList combined;
+            size_t combined_allocated = 0;
+
+            auto filter_holder_by_selector = [&](const HashJoin::NullMapHolder & holder)
+            {
+                const auto * sc = holder.columns;
+                if (!sc)
+                    return; // skip broken holder
```

Comment:
What does "broken" mean here? We shouldn't have such places in code, where we expect something to be broken and just skip it over.